backend/tarsy/agents/iteration_controllers/base_controller.py

```
@@ -253,14 +259,30 @@ async def execute_analysis_loop(self, context: 'StageContext') -> str:
                 error_continuation = self.parser.get_error_continuation(str(e))
                 conversation.append_observation(error_continuation)
                 continue
                 
-        # 8. Timeout handling
-        self.logger.warning("ReAct analysis reached maximum iterations without final answer")
-        return f"Analysis incomplete: reached maximum iterations ({max_iterations}) without final answer"
+        # 8. Timeout handling - check if stage should be marked as failed
+        if last_interaction_failed:
+            # Stage failure: reached max iterations with failed last interaction
+            from ..exceptions import MaxIterationsFailureError
+            self.logger.error(f"Stage failed: reached maximum iterations ({max_iterations}) with failed last interaction")
```

Ruff (0.12.2)

Line too long (122 > 88)

(E501)